### ISSUE TITLE
After Scenario Method in RP class to modify things in Report Portal

### ISSUE THREAD

**Issue by Shivin89** [kirksl/karate-maven-gradle#13]
Hi @kirksl 

Can we have the After Scenario Method in the RP class, I want to modify some things at the Scenario level in Report Portal,  I know we have an after scenario in Hook Class, but wanted an After Scenario like we have after feature at Report Portal class. 

Sometimes users might want to change attributes or like for me I want to integrate with  Zephyr  (I have already modified at launch level and other details, but I want to get test case id from scenario (@testcaseid=xxxx) and XXXX would be zephyr ticket which will update the status of that id ) I have implemented this in Report Portal Cucumber plugin and it works pretty well.

Cucumber plugin  after scenario (Modified to update Zephyr with that test case)
protected void afterScenario(TestCaseFinished event) throws JSONException, IOException, InterruptedException {
		
		RunningContext.ScenarioContext context = getCurrentScenarioContext();
		URI featureUri = context.getFeatureUri();
		currentScenarioContextMap.remove(Pair.of(context.getLine(), featureUri));
		Date endTime = finishTestItem(context.getId(), event.getResult().getStatus());		
		featureEndTime.put(featureUri, endTime);
		
		Collection<String> tags = event.getTestCase().getTags();
		if (tags.isEmpty()) {
			testCaseId = null;
		}
		for (String i : tags) {
			testCaseId = i;
			if ((testCaseId).contains("@testcase=")) {
				testCaseId = testCaseId.substring("@testcase=".length());
				break;
			} else {
				testCaseId = null;
			}
		}		

		if (testCaseId != null && launcName!=null)
			try {
				{
					ZapiIntegration zapi = ZapiIntegration.getInstance();
					zapi.UpdateZephyrStatus(testCaseId, event.getResult().getStatus().toString());

				}
			} catch (Exception e) {
				
				e.printStackTrace();
			}
		else {
			System.out.println("TestCase id is not added to Scenario in feature file, or Zephyr Test Cycle is not provided");
			System.out.println("Zephyr Test execution will not get updated: Please update Test case id or add Test Cycle");
		}
		
		currentScenarioContext.set(null);
		removeFromTree(currentFeatureContextMap.get(context.getFeatureUri()), context);
	}

Please suggest what you think about this?

**Comment by kirksl** [comment]
Hi @Shivin89.  Apologies for the delay i've been very busy.  

We initially tried to use "after scenario" but had many issues when running tests in parallel.  The results posted to RP were reported incorrectly.  I spent a lot of time on this to see what others were doing and understand RP code.  Ultimately the safest option and what I saw others doing was to take advantage of Java "synchronize" and to report results using "after feature" where we could still access all of the scenario level detail, but it would get reported all at once per feature.

I would suggest keeping what is in place as you have full capability to spin through the scenarios in the "after feature" hook.  If you really need to use "after scenario" feel free to try, but be very careful, test parallel execution, verify results are always correct.  The other issue we saw sporadically when using "after scenario" is that execution runs would not complete as well and we had to force finish test runs in the RP UI.

**Comment by Shivin89** [comment]
Hi @kirksl 

Thanks for the explanation, appreciate it !!  I had one question not related to this issue but in general, Is the Calling Java class working in your framework?  like example below 

def JavaDemo = Java.type('com.cdk.runner.RPReporter').SL_DC
* json result = JavaDemo.getSlDc("HelloWorld")
  * print JavaDemo 

https://stackoverflow.com/questions/70283234/set-the-variable-value-defined-in-the-java-class-via-feature-file

Thanks
Shivin

**Comment by kirksl** [comment]
@Shivin89 are you having trouble with a specific Java class or calling all Java classes?  I am able to call Java classes within karate-maven-gradle.

**Comment by Shivin89** [comment]
Hi @kirksl  I was trying to call Report Porter Class using 
def JavaDemo = Java.type('com.cdk.runner.RPReporter')

I have defined the below variables in the Report Portal class and 
public static  String SL_JOB_ID;
public static String SL_DC = "US";

I have this method (Setter and Getter)
public  void setSlJobId(String slJobId) {

        SL_JOB_ID = slJobId;
}


    public static String getSlDc() {
        return SL_DC;
    }

I am trying to set the value of SL_JOB_ID to session-id in my feature file 

(
 Given driver baseUrl
    * def sessionId = driver.sessionId
    * print sessionId
 def JavaDemo = Java.type('com.cdk.runner.RPReporter')
    * def test =  JavaDemo.setSlJobId("Testing") 

But when I run the test it says 
01: JavaDemo.setSlJobId("Testing")
<<<<
org.graalvm.polyglot.PolyglotException: TypeError: invokeMember (setSlJobId) on com.cdk.runner.RPReporter failed due to: Unknown identifier: setSlJobId
- <js>.:program(Unnamed:1)

I have also tried to get the value of  SL_DC too but it also failed with the same error 

I have also tried def JavaDemo = Java.type('com.cdk.runner.RPReporter').SL_DC  and test passes but the value of JavaDemo  is Null when I print it
-----------------
I am able to call a library called Faker and use methods inside that library 
in karate-config.js
 config.faker = Java.type('com.github.javafaker.Faker');
 * def fakerObj =  new faker()
    * def fName = fakerObj.name().firstName()

And this works fine , I have tried creating obj of the class too but then it failed saying message not supported .

This is a static method so we don't need to create object .

**Comment by kirksl** [comment]
@Shivin89 this thread is going out of scope of the issue reported.  

But below is an example of how I use  local/project classes if that helps.  I'd suggest talking to Karate developers if you can't solve this.  Also try other versions of Karate maybe it's a bug?

karate-config.js
```
var dbService = Java.type('common.DBService');
config.db = new dbService();
return config;

// now db class and methods are available everywhere
```

**Comment by kirksl** [comment]
Closing due to inactivity
